File: backend/app/schemas/hujjat.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict

from app.models.audit_log import AuditAmal
from app.models.foydalanuvchi import Smena
from app.models.kip import KipHolati


class HujjatKipJavob(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    mahsulot_kodi: str
    mahsulot_nomi: str
    partiya_id: int
    partiya_raqami: int
    kip_raqami: int
    ogirlik: float
    smena: Smena
    operator_id: int
    operator_ism: str
    vaqt: datetime
    surat_yoli: str | None
    holati: KipHolati


class DavrOraligiJavob(BaseModel):
    """Hujjatlar ekranidagi "Kunlik/Haftalik/Oylik/Mavsum" tezkor davr
    tugmalari uchun — `sana_dan`/`sana_gacha`ni frontend o'zi hisoblamasdan,
    backenddagi `davr_oraligi()` (mavsum_boshlanish_sanasi sozlamasidan
    foydalanadigan) mantig'idan oladi — Statistika/Dashboard bilan izchil."""

    davr: str
    sana_dan: date
    sana_gacha: date


class AuditLogJavob(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    foydalanuvchi_id: int
    foydalanuvchi_ism: str
    jadval_nomi: str
    yozuv_id: int
    amal: AuditAmal
    eski_qiymat: dict | None
    yangi_qiymat: dict | None
    sabab: str
    vaqt: datetime

File: frontend/lib/models/hujjat.dart
class HujjatKip {
  final int id;
  final String mahsulotKodi;
  final String mahsulotNomi;
  final int partiyaRaqami;
  final int kipRaqami;
  final double ogirlik;
  final String smena;
  final String operatorIsm;
  final DateTime vaqt;
  final String? suratYoli;
  final String holati;

  HujjatKip({
    required this.id,
    required this.mahsulotKodi,
    required this.mahsulotNomi,
    required this.partiyaRaqami,
    required this.kipRaqami,
    required this.ogirlik,
    required this.smena,
    required this.operatorIsm,
    required this.vaqt,
    this.suratYoli,
    required this.holati,
  });

  factory HujjatKip.fromJson(Map<String, dynamic> j) => HujjatKip(
        id: j['id'],
        mahsulotKodi: j['mahsulot_kodi'],
        mahsulotNomi: j['mahsulot_nomi'],
        partiyaRaqami: j['partiya_raqami'],
        kipRaqami: j['kip_raqami'],
        ogirlik: (j['ogirlik'] as num).toDouble(),
        smena: j['smena'],
        operatorIsm: j['operator_ism'],
        vaqt: DateTime.parse(j['vaqt']),
        suratYoli: j['surat_yoli'],
        holati: j['holati'],
      );
}

class Sahifalangan<T> {
  final List<T> items;
  final int jami;
  final int sahifa;
  final int sahifaHajmi;

  Sahifalangan({required this.items, required this.jami, required this.sahifa, required this.sahifaHajmi});

  factory Sahifalangan.fromJson(Map<String, dynamic> j, T Function(Map<String, dynamic>) itemFromJson) =>
      Sahifalangan(
        items: (j['items'] as List).map((e) => itemFromJson(e)).toList(),
        jami: j['jami'],
        sahifa: j['sahifa'],
        sahifaHajmi: j['sahifa_hajmi'],
      );
}
